### PR TITLE
Criteo DLRM PyTorch Implementation

### PR DESCRIPTION
**Pull request by pomonam**
Many codes are directly copied from PR #89.

- Added Pytorch implementation for Criteo1tb (following the Jax code). 
- Unified input pipelines implemented and forward pass working (with a single GPU). 
- Unit tests for model comparison implemented and confirmed its passing. 

TODOs:
- I am working with @rakshithvasudev to see if it meets the target accuracy with JAX. 
- Need to see if the implementation works with multi-GPU.

### CODE REVIEW

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    np_iter = super().build_input_queue(data_rng,
+                                        split,
+                                        data_dir,
+                                        global_batch_size,
+                                        num_batches,
+                                        repeat_final_dataset)
+    while True:
+      # Only iterate over tf input pipeline in one Python process to
+      # avoid creating too many threads.
+      if RANK == 0:
```
Comment (pomonam) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:129]: @runame It would be helpful if you can take a look at this part; I borrowed the code from WMT (with the modification that I think is needed). To give you more background, we have 'inputs', 'targets', and 'weights' in the batch and the dimension would be `batch_size, 39`. Thanks a lot!

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/models.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        nn.init.normal_(
+            module.weight.data,
+            0.,
+            math.sqrt(2. / (module.in_features + module.out_features)))
+        nn.init.normal_(module.bias.data,
+                        0.,
+                        math.sqrt(1. / module.out_features))
+
+  def forward(self, x):
+    bot_mlp_input, cat_features = \
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/models.py:121]: nit: different style for consistency
```python
torch.split(
    x, [self.num_dense_features, self.num_sparse_features], 1)
```

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        batch,
+        model_state=None,
+        mode=spec.ForwardPassMode.EVAL,
+        rng=None,
+        update_batch_norm=False)
+    per_example_losses = metrics.per_example_sigmoid_binary_cross_entropy(
+        logits, batch['targets'])
+    batch_loss_numerator = torch.sum(per_example_losses)
+    batch_loss_denominator = np.sum(batch['weights'])
+    return np.asarray(batch_loss_numerator), batch_loss_denominator
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:177]: I'm probably missing something, but I think `np.asarray()` cannot be called on a GPU tensor?

File: tests/test_num_params.py [hunk 4]
```diff
@@ -76,6 +81,28 @@ def get_models(workload):
...
+    input_shape = (1, 39)
+    model_init = JaxDlrmSmall(
+        vocab_sizes=tuple([1024 * 128] * 26),
+        total_vocab_sizes=sum(tuple([1024 * 128] * 26)),
+        num_dense_features=13,
+        mlp_bottom_dims=(128, 128),
+        mlp_top_dims=(256, 128, 1),
+        embed_dim=64).init
+    jax_model = model_init(init_rngs, jnp.ones(input_shape, jnp.float32),
+                           False)["params"]
```
Comment (runame) [inline tests/test_num_params.py:96]: nit: single quotation marks for consistency

**Review by runame — COMMENTED** [review]
I have taken a quick look and left a few comments/questions. To verify that it works correctly with DDP, I would have to try running it myself.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                                        repeat_final_dataset)
+    while True:
+      # Only iterate over tf input pipeline in one Python process to
+      # avoid creating too many threads.
+      if RANK == 0:
+        batch = next(np_iter)  # pylint: disable=stop-iteration-return
+        tensor_list = []
+        for key, value in batch.items():
+          tensor = torch.as_tensor(value, dtype=torch.float32, device=DEVICE)
+          tensor_list.append(tensor)
```
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:134]: Ok, you have actually already told me that this is the case :D

**Comment by rakshithvasudev** [comment]
Yes we're working on this right now. Will keep you posted :) 


Thanks!

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 6]
```diff
@@ -117,24 +117,26 @@ def build_input_queue(self,
 
     # The input pipeline has to be created in all processes, because
     # self._tokenizer has to be available in every process.
```
[the remark this answers is not in the document]
Comment (pomonam) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py]: Fixed!

**Comment by pomonam** [comment]
@runame Update: we hit the initial loss with the same order of magnitude as the JAX code and confirmed that the loss increases. 
There seems to be a problem with multi-GPU, @rakshithvasudev reported deadlock. We need to look into the deadlock issue.

**Comment by runame** [comment]
Thanks for the update! At what point does the deadlock occur?

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                        data_dir: str,
+                        global_batch_size: int,
+                        num_batches: Optional[int] = None,
+                        repeat_final_dataset: bool = False):
+    per_device_batch_size = int(global_batch_size / N_GPUS)
+
+    # Only create and iterate over tf input pipeline in one Python process to
+    # avoid creating too many threads.
+    if RANK == 0:
+      data_rng = data_rng.astype('uint32')
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:120]: Nit: this line is not necessary (data_rng will be deleted in `super().build_input_queue()`).

**Comment by runame** [comment]
In the input pipeline you first batch using the `per_device_batch_size` and then `num_devices` ([here](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py#L91) and [here](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py#L99)). I'm not sure about the behavior of this function, but shouldn't we first batch using the `global_batch_size`? Then we could also use the `data_utils.shard_numpy_ds` afterwards, as for [WMT](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/wmt/input_pipeline.py#L299) (or probably also batch using `num_devices`, like it is now).

File: algorithmic_efficiency/workloads/criteo1tb/workload.py [hunk 8]
```diff
@@ -85,9 +85,17 @@ def build_input_queue(self,
         vocab_sizes=self.vocab_sizes,
         num_batches=num_batches,
         repeat_final_dataset=repeat_final_dataset)
-    for batch in iter(ds):
-      batch = jax.tree_map(lambda x: x._numpy(), batch)  # pylint: disable=protected-access
-      yield batch
+
+    # for batch in iter(ds):
+    #   batch = jax.tree_map(lambda x: x._numpy(), batch)  # pylint: disable=protected-access
+    #   yield batch
```
Comment (znado) [inline algorithmic_efficiency/workloads/criteo1tb/workload.py:91]: I'd just delete these

File: algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py [hunk 9]
```diff
@@ -88,14 +100,16 @@ def _parse_example_fn(example):
...
-  ds = ds.batch(per_device_batch_size, drop_remainder=is_training)
-  ds = ds.map(_parse_example_fn, num_parallel_calls=16)
+  ds = ds.batch(global_batch_size, drop_remainder=is_training)
+  ds = ds.map(_parse_example_fn, num_parallel_calls=AUTOTUNE)
+  ds = ds.prefetch(tf.data.AUTOTUNE)
+
   if num_batches is not None:
     ds = ds.take(num_batches)
-  # We do not need a ds.cache() because we will do this anyways with
-  # itertools.cycle in the base workload.
```
Comment (znado) [inline algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py:96]: can we keep this comment?

File: tests/test_num_params.py [hunk 10]
```diff
@@ -76,6 +81,28 @@ def get_models(workload):
...
+  elif workload == 'criteo1tb':
+    # Init Jax model.
+    input_shape = (1, 39)
+    model_init = JaxDlrmSmall(
+        vocab_sizes=tuple([1024 * 128] * 26),
+        total_vocab_sizes=sum(tuple([1024 * 128] * 26)),
+        num_dense_features=13,
+        mlp_bottom_dims=(128, 128),
+        mlp_top_dims=(256, 128, 1),
+        embed_dim=64).init
```
Comment (znado) [inline tests/test_num_params.py:94]: can we make these kwargs a dict and pass them via **kwargs to both jax and pytorch versions?

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                        split: str,
+                        data_dir: str,
+                        global_batch_size: int,
+                        num_batches: Optional[int] = None,
+                        repeat_final_dataset: bool = False):
+    per_device_batch_size = int(global_batch_size / N_GPUS)
+
+    # The input pipeline has to be created in all processes, because
+    # self._tokenizer has to be available in every process.
+    np_iter = super().build_input_queue(data_rng,
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:126]: assuming we don't need to include the `RANK == 0` check, lets delete the commented out code? if we wanna test it for debugging we can go back and grab it?

**Review by znado — APPROVED** [review]
looks fantastic, thanks so much for the quick turn around on this!

**Comment by pomonam** [comment]
@znado Thanks for the comments! The current version has a lot of commented out stuff as it was under development. I will clean it in the next revision. 

@rakshithvasudev has some good news to share. On a single GPU, we reach the target val loss with this version of the code. Regarding, the multi-gpu problem, on @rakshithvasudev's hardware, he is also getting the same (deadlock?) issue on ogbg workload, which uses the unified pipeline. @runame, did you previously observe any issue on ogbg workload?

**Comment by rakshithvasudev** [comment]
Here is the error for OGBG : https://github.com/mlcommons/algorithmic-efficiency/issues/149

It would be worth it to verify it criteo pytorch runs with 8 GPUs in your env @runame, we know ogbg ran without any issues in your env.

**Comment by runame** [comment]
Sure, I'll try to run it on my setup. Where can I get the data for the Criteo workload?

**Comment by rakshithvasudev** [comment]
Great! Thanks.

Here is the link to dataset:

https://ailab.criteo.com/download-criteo-1tb-click-logs-dataset/

Once you download and extract it, you want to use the Linux utility “split”
on these extracted files in create multiple files where each file has 5
million lines.

These files are needed by the tf data pipeline.

Thanks
Rakshith

**Comment by runame** [comment]
Using the wrong shape when initializing the variable to receive a broadcasted tensor can actually result in a deadlock (which is why I asked about it [here](https://github.com/mlcommons/algorithmic-efficiency/pull/146#discussion_r982925493) ;)), did fixing it resolve the issue? Didn't get to try running it on my setup yet, but have downloaded the data.

**Comment by pomonam** [comment]
@runame Thank you! Yes, we just fixed it and was running the code. It does not have a deadlock issue anymore. @rakshithvasudev can maybe update us to see if the target loss is achieved with multi-gpu.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pomonam) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py]: Thanks! Yes, they had different shapes and I fixed this issue on my latest commit.

**Comment by runame** [comment]
Awesome! Let me know if any new issues with DDP occur.

**Comment by rakshithvasudev** [comment]
Training with multiGPU, loss is improving!

```
I1003 21:30:55.437373 140737350199104 submission_runner.py:295] 26.69s  1       {'train/loss': tensor(0.1785, device='cuda:0'), 'validation/loss': tensor(0.1912, device='cuda:0'), 'test/loss': tensor(0.1939, device='cuda:0')}
I1003 21:30:56.239438 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.240317 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.241438 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.242493 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.242469 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.242682 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.242746 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:30:56.242919 140737350199104 distributed.py:977] Reducer buckets have been rebuilt in this iteration.
I1003 21:37:40.322416 140737350199104 spec.py:258] Evaluating on the training split.
I1003 21:37:48.770949 140737350199104 spec.py:270] Evaluating on the validation split.
I1003 21:39:10.507946 140737350199104 spec.py:285] Evaluating on the test split.
I1003 21:40:33.161107 140737350199104 submission_runner.py:295] 627.14s         143     {'train/loss': tensor(0.1131, device='cuda:0'), 'validation/loss': tensor(0.1299, device='cuda:0'), 'test/loss': tensor(0.1370, device='cuda:0')}
```

**Comment by rakshithvasudev** [comment]
I can confirm this code trains with 8 GPUs without deadlocks.

File: algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py [hunk 12]
```diff
@@ -9,9 +9,16 @@
...
 from typing import Optional, Sequence
 
-import jax
 import tensorflow as tf
 
+from algorithmic_efficiency import data_utils
+from algorithmic_efficiency.pytorch_utils import pytorch_setup
+
+RANK = pytorch_setup()[1]
+AUTOTUNE = tf.data.AUTOTUNE if RANK == 0 else None
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py:18]: This condition is not necessary here since the input pipeline is only created in the process with `RANK == 0`.

File: algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py [hunk 13]
```diff
@@ -80,6 +87,12 @@ def _parse_example_fn(example):
     num_files = math.ceil(num_examples / _CSV_LINES_PER_FILE)
     ds = ds.take(num_files)
 
+  # Avoid creating too many threads when using PyTorch DDP.
+  if RANK != 0:
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/input_pipeline.py:91]: See comment above, can be removed.

File: algorithmic_efficiency/workloads/criteo1tb/workload.py [hunk 14]
```diff
@@ -85,9 +88,14 @@ def build_input_queue(self,
         vocab_sizes=self.vocab_sizes,
         num_batches=num_batches,
         repeat_final_dataset=repeat_final_dataset)
-    for batch in iter(ds):
-      batch = jax.tree_map(lambda x: x._numpy(), batch)  # pylint: disable=protected-access
-      yield batch
+
+    # Separate function is necessary because the code above has to be executed
+    # when build_input_queue is called (not when next() is first called on it).
+    def _input_queue_generator():
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/workload.py:94]: In this case no separate function is necessary. This was only needed for WMT because the tokenizer had to be created in all processes.

File: submission_runner.py [hunk 15]
```diff
@@ -315,6 +315,7 @@ def train_once(
   metrics = {'eval_results': eval_results, 'global_step': global_step}
   if USE_PYTORCH_DDP:
     # Sync final score (accumulated training time); choose highest, i.e. worst.
+    dist.barrier()
```
Comment (runame) [inline submission_runner.py:318]: It surely doesn't hurt, but out of curiosity, did you see any issue without the barrier here?

**Review by runame — APPROVED** [review]
Great to hear that everything seems to work now. I have left a few last comments. After they are addressed the PR can be merged, unless @znado has anything else.

File: submission_runner.py (hunk 15 above)
Comment (pomonam) [reply submission_runner.py:318]: @rakshithvasudev had an issue with it and was solved after adding a barrier here.

Comment (rakshithvasudev) [reply submission_runner.py:318]: Is without that the summary was deadlocked :)

Comment (runame) [reply submission_runner.py:318]: Ok interesting, thanks for fixing it!

**Comment by pomonam** [comment]
@runame Thanks for the comments again. The new commit should address the above issues.

**Comment by runame** [comment]
Perfect, thanks!

**Comment by znado** [comment]
looks fantastic to me!
